### ISSUE TITLE
Inline bind operator for computation expressions

### ISSUE THREAD

**Issue by jwosty** [fsharp/fslang-suggestions#1070]
# Title of Suggestion

I propose we add an operator or keyword which can be used to bind values in computation expressions deep inside expressions (potentially arbitrarily), similar to C#'s await keyword, which can be used in the middle of expressions, and is not limited to top-level expressions in the computation expression. The specific syntax is of lesser issue than the form of these expressions themselves.

For sake of discussion, I am going to use a ``bind!`` syntax in this proposal (which could easily be replaced with one of the alternative choices).

The following code:

```fsharp
if (f(bind! x) && bind! y) then // ...
```

would be lifted to something equivalent to:

```fsharp
let! x' = x
let! y' = y
if (f x && y) then // ...
```

There are many more cases where this should likely work, such as:

```fsharp
// right-hand of let-expr
let x = bind! expr
// deep inside right-hand of let-expr
let x = (bind! expr) |> List.map string
// match expr
match bind! expr with
```

And so on -- there are likely more cases to consider and decide on.

However, this would not be allowed arbitrarily deep, as the following should obviously be invalid:

```fsharp
async {
    let f = (fun () -> bind! expr) // compiler error -- not allowed inside lambdas
    let g x =
        bind! expr // compiler error -- not allowed inside nested functions
}
```

There are also probably other exclusions that should be fleshed out.

The current way to approach this is to manually lift the bindable expression and bind it using ``let!``.

## Pros and Cons 

The advantages of making this adjustment to F# are:

* more parity with C#'s deeply nestable ``await`` keyword
* highly useful because it allows the elimination of many otherwise unnecessary variable names in source code

The disadvantages of making this adjustment to F# are:

* more complexity

## Extra information

Estimated cost (XS, S, M, L, XL, XXL): L (I'm guessing)

Related suggestions: (put links to related suggestions here)

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [x] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [x] This is not a breaking change to the F# language design
* [x] I or my company would be willing to help implement and/or test this

## For Readers

If you would like to see this issue implemented, please click the :+1: emoji on this issue. These counts are used to generally order the suggestions by engagement.

**Comment by jwosty** [comment]
Which syntax would you most prefer (vote using reactions)?

I believe that ``bind`` and ``await`` (without the bang) would both be backward-incompatible, so I have left them out of this vote.

* ``bind!`` - 👍  
* ``await!`` - ❤️ 
* postfix ``!`` - 🚀

**Comment by baronfel** [comment]
As an implementation detail, I think that the amount of binds in a complex expression could be reduced by using the `let!/and!` form if the CE builder supported it, because in the expression `if f(a!) && b! then ... else ...`, a's bind and b's bind are independent operations. I'm sure there are other opportunities for this kind of expansion as well, we've just been focusing on `if` here.

**Comment by jwosty** [comment]
Advantages of each potential operator/keyword:

* ``bind!``
  * Same name as the CE Bind method, and gives hints to their monadic heritage
  * Has a symmetry with ``let!`` in that the bang comes before the expression it is binding
* ``await!``
  * Similar to C# and Javascript's async/await syntax -- would be a more familiar experience for C# devs learning F#
  * Symmetry with ``let!``
* postfix ``!``
  * Less verbose

**Comment by dsyme** [comment]
> Which syntax would you most prefer (vote using reactions)?

We would likely make it CE-configurable.  But votes on preferred syntax for `async`, `task`, `option`, `asyncOption`, `asyncSeq` etc. would be welcome.

**Comment by jwosty** [comment]
> 
> 
> > Which syntax would you most prefer (vote using reactions)?
> 
> We would likely make it CE-configurable. But votes on preferred syntax for `async`, `task`, `option`, `asyncOption`, `asyncSeq` etc. would be welcome.

So more like a custom operation? Sounds interesting. How are you envisioning that to work, specifically?

**Comment by dsyme** [comment]
> So more like a custom operation? Sounds interesting. How are you envisioning that to work, specifically?

Yes I'd imagine either a new method or a new attribute on the CE builder.  I've not thought about it more than that though.

**Comment by ronaldschlenker** [comment]
Related to #1000

**Comment by jl0pd** [comment]
It would be awesome if `bind` could go beyond binding single value:
```fsharp
let await2Tasks t1 t2 = // Task<'a> -> Task<'b> -> Task<('a * 'b)>
    task {
        let (r1, r2) = 
             bind t1
             and t2
        return (r1, r2)
    }
// which is identical to
let await2Tasks t1 t2 = // Task<'a> -> Task<'b> -> Task<('a * 'b)>
    task { 
       let! _ = Task.WhenAll(t1, t2)
       return (t1.Result, t2.Result) 
    }
```

Right hand side of `bind`  and `and` is any expression that can be bound with `let!`
```fsharp
let get42Async () =
    task {
          let (four, ten, two) =
                bind Task.Run (fun () -> 4)
                and Task.Run (fun () -> 10)
                and Task.Run (fun () -> 2)
          return four * ten + two
    }
```

**Comment by Tarmil** [comment]
@jl0pd I don't think this would need anything extra if there's a bind operator, you could just do this:

```fsharp
task {
    let (four, ten, two) =
        (bind Task.Run(fun () -> 4),
         bind Task.Run(fun () -> 10),
         bind Task.Run(fun () -> 2))
    return four * ten + two
}
```

What I wonder, though, is whether the above should automatically use applicative methods (MergeSources / BindReturn) if they are defined, or be restricted to Bind.

**Comment by lambdakris** [comment]
Hmm, @Tarmil and @jl0pd, would it make sense to use tailing `bind`'s to invoke `Bind` and use tailing `and`'s to invoke applicative methods (`MergeSources`/`BindReturn`)? Being a bit of a novice with computation expressions, I wonder if that would be: 
1. A useful capability to have? 
2. Congruent with the current behavior of applicative expressions?

**Comment by jl0pd** [comment]
> What I wonder, though, is whether the above should automatically use applicative methods (MergeSources / BindReturn) if they are defined, or be restricted to Bind.

@Tarmil, it should not automatically use `MergeSources` because developer may rely on sequential execution, e.g. `createUser` and then `setAdditinalInfo`. If last is going to execute before first (which may happen with `Task.WhenAll`) then it may fail

**Comment by kspeakman** [comment]
Love this suggestion. Found it after working with task CEs and feeling the need for new bang syntax like `while!` and `if!`. Then thought rather than making a bang version of everything, why not have 1 keyword that semantically unwraps the value like `await` does in C#? (Recognizing that it's no less complicated to implement behind the scenes.)

Re names. "bind" is an obstacle for newcomers. As a vague/archaic word, it harms readability in a similar way to using symbols. And its meaning has more to do with its implementation than its apparent usage. "await" is fantastic for readability of asyncs, less for other things. A more general fit for the user's intent might be "unwrap". Postfix bang `x!` is subtle / easy to miss. Also consider multi-term expressions:

```fsharp
while (channel.Reader.WaitToReadAsync())! do ...

while await! channel.Reader.WaitToReadAsync() do ...
```

@jwosty The side-by-side examples seem to use extra parenthesis: `if (await! x) && y then ...`. Do CE keywords have/follow order of precedence rules? If so, `if await! x && y then ...` should work.

**Comment by jwosty** [comment]
> Love this suggestion. Found it after working with task CEs and feeling the need for new bang syntax like `while!` and `if!`. Then thought rather than making a bang version of everything, why not have 1 keyword that semantically unwraps the value like `await` does in C#? (Recognizing that it's no less complicated to implement behind the scenes.)

Agreed, this proposal could obsolete all the special-case bang keywords

> @jwosty The side-by-side examples seem to use extra parenthesis: `if (await! x) && y then ...`. Do CE keywords have/follow order of precedence rules? If so, `if await! x && y then ...` should work.

That might be true; I just included them for clarity (I tend to overuse parenthesis :) )

**Comment by SchlenkR** [comment]
This is _the_ most wanted feature of me. It would simplify the code I write for some libraries a lot.

Example: Signal processing, where parameters have to be modulated often:

```fsharp
let! mod = Osc.sine(frq = 120.0)
return! Osc.rect(frq = 4000.0 * mod)

// ...with inline bind (here using a prefix op "!!" just for demo)
return! Osc.rect(frq = 4000.0 * (!!Osc.sine(frq = 120.0)))
```

Another example: Arithmetic operations, which would reduce the need for custom operators and / or SRTP overload "hacks". The advantage here is the inline-style of writing the operands as one would expect it in an equation.

```fsharp
// nice: we can use "+" as we would expect it to be used :)
let result1 = !!Osc.square(frq = 100.0) + !!Osc.sine(200.0)
let result2 = 100.0 + !!Osc.sine(200.0)

// current alt. 1:
let! a = Osc.square(frq = 100.0)
let! b = Osc.sine(200.0)
let result1 = a + b

// current alt. 1 using custom op (here: ".+", ".+." or "+."), which can a pain regarding all the complexity that numbers introduce
let result1 = Osc.square(frq = 100.0) .+. Osc.sine(200.0)
let result2 = 100.0 +. Osc.sine(200.0)
// ...
```

Currently, I am currently developing a UI library that works like react, and uses only Skia under the hood, where one can define triggers, animations, whatever kind of state-preserving functions. Using animations inline, e.g. for the x-coord of a UI element, would reduce the brain-load, make it so nice to write just in one-shot. That would be great.

I like how C# async is being able to be inlined; it's a quite seamless integration and feels really natural. I don't care if it's an operator or a keyword, but usable as simple as inlined async in C#.

If this could be defined, I would perhaps dare to implement it. cc @vzarytovskii @dsyme  ?

**Comment by dsyme** [comment]
I'm marking this as approved-in-principle as a language design item.  I think `!!` would be appropriate.

Implementing it requires a significant amount of work.

**Comment by OnurGumus** [comment]
> I'm marking this as approved-in-principle as a language design item. I think `!!` would be appropriate.
> 
> Implementing it requires a significant amount of work.

Please note that, ``!!`` is somewhat common in Fable code which is shorthand for unbox<_>

**Comment by brianrourkeboll** [comment]
https://github.com/fsharp/fslang-suggestions/issues/1070#issuecomment-912746162

> > Which syntax would you most prefer (vote using reactions)?
> 
> We would likely make it CE-configurable. But votes on preferred syntax for `async`, `task`, `option`, `asyncOption`, `asyncSeq` etc. would be welcome.

https://github.com/fsharp/fslang-suggestions/issues/1070#issuecomment-912750331

> > So more like a custom operation? Sounds interesting. How are you envisioning that to work, specifically?
> 
> Yes I'd imagine either a new method or a new attribute on the CE builder. I've not thought about it more than that though.

So something like a `CustomKeywordAttribute` taking a custom name that could be applied to methods on the computation expression builder (including `Bind`, etc.)? In usage, I assume the custom name would always need to be followed by `!`.

```fsharp
open System

[<Sealed>]
[<AttributeUsage(AttributeTargets.Method, AllowMultiple=false)>]
type CustomKeywordAttribute (name : string) =
    inherit Attribute ()
    member _.Name = name
```

Since identifiers ending with `!` are currently disallowed, this would be fully backwards-compatible. This would allow domain-specific names to be used for different computation expressions, like, say, `await!` for things like `task` and `async`, but other names for other usages.

```fsharp
type TaskBuilder () =
    [<CustomKeyword("await")>]
    member _.Bind (task, continuation) = …
```

```fsharp
let f g x y z =
    task {
        if await! (g x) then
            return y
        else
            return z
    }
```

```fsharp
let f x y z =
    validation {
        return
            { A = apply! X.tryParse x
              B = apply! Y.tryParse y
              C = apply! Z.tryParse z }
    }
```

I assume that we'd need to allow applying `CustomKeywordAttribute` with the same name to multiple overloads.

**Comment by gusty** [comment]
I think that `!` should do the opposite when used on a right side expression, I mean it should lift/wrap the value, possibly by calling `Return`.

This way we will have:

```
let! x = x!
```
equivalent to
```
let x = x
```

So, they cancel each other.
This will address another potential suggestion (as I always miss a shortcut for this operation) by providing a nice symmetric approach.

We still have other operators available for this suggested bind operation, like `^` or `$` to name the first two that come to my mind.

**Comment by jwosty** [comment]
> I think that `!` should do the opposite when used on a right side expression, I mean it should lift/wrap the value, possibly by calling `Return`.

Interesting idea, but personally, I'm not sure I'm sold since you can always just do `ce { return x }` (it's not _that_ much more verbose for how often at least I come across it)
